Metrics: make sure that the duration is always bigger than 0.

eloycoto opened the pull request:
Found today in a user instance that the endpoint regeneration was
negative, this commit will make sure to not update invalid data to
metrics.

```
cilium_endpoint_regeneration_time_stats_seconds_sum{scope="waitingForLock",status="success"} -3.6893488146287866e+11
```

Review the code before doing this simple fix, and here is my conclusion: 
- The snapStat package is ok, tested all options and I didn't find the issue in there. 
- Review all context in `RegenerateBPF` and all looks good to me. 
- Review `observe` code and was ok too. I also did some monkey patching around prometheus. 

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5889)
<!-- Reviewable:end -->

joestringer reviewed (changes requested):
Is there any legitimate reason for these durations to be negative?

Perhaps we should at least raise some kind of log if they are negative so we can at least try to investigate further? I'm a bit worried that this will just cause us to lose some visibility and have no idea why.

eloycoto committed:
Spanstat: Validate that the duration is not negative when call End.

To make sure that no negative values are added in the snapstat validate
that duration is bigger or equal than 0.

Also, if the number is negative will print the stacktrace in the stderr
output.

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

tgraf commented on pkg/spanstat/spanstat.go:
```diff
@@ -37,10 +45,14 @@ func (s *SpanStat) Start() {
+			}
 		} else {
-			s.failureDuration += d
+			log.WithField("duration", d).Warn("Duration is negative on End")
```
We need a stack trace for this to be useful

eloycoto replied:
Fixed.

ianvernon commented on pkg/spanstat/spanstat.go:
```diff
@@ -15,7 +15,16 @@
 package spanstat
```
commit uses incorrect package name: snapstat --> spanstat, both in title and commit message.

ianvernon commented on pkg/spanstat/spanstat.go:
```diff
@@ -15,7 +15,16 @@
+)
+
+var (
+	subSystem = "snapstat"
```
snapstat --> spanstat

joestringer commented on pkg/endpoint/metrics.go:
```diff
@@ -65,10 +66,10 @@ func (s *regenerationStatistics) SendMetrics() {
 			metrics.EndpointRegenerationTimeStats.WithLabelValues(scope, "success").Observe(stat.SuccessTotal().Seconds())
 		}
-		if stat.FailureTotal() != time.Duration(0) {
+		if stat.FailureTotal() > time.Duration(0) {
```
Do we still need these changes?

Logically if somehow a spanstat doesn't get `End()`ed due to another bug, then presumably this would still silently discard metrics?

joestringer reviewed (approved):
LGTM, thanks!
